pkg/pod/status.go

```diff
@@ -224,22 +232,36 @@ func createMessageFromResults(results []v1beta1.PipelineResourceResult) (string,
 	return string(bytes), nil
 }
 
-func filterResultsAndResources(results []v1beta1.PipelineResourceResult) ([]v1beta1.TaskRunResult, []v1beta1.PipelineResourceResult, []v1beta1.PipelineResourceResult) {
+func filterResultsAndResources(results []v1beta1.PipelineResourceResult, specResults []v1beta1.TaskResult) ([]v1beta1.TaskRunResult, []v1beta1.PipelineResourceResult, []v1beta1.PipelineResourceResult) {
 	var taskResults []v1beta1.TaskRunResult
 	var pipelineResourceResults []v1beta1.PipelineResourceResult
 	var filteredResults []v1beta1.PipelineResourceResult
+	neededTypes := make(map[string]string)
+	// collect needed types for results
+	for _, r := range specResults {
+		neededTypes[r.Name] = string(r.Type)
+	}
 	for _, r := range results {
 		switch r.ResultType {
 		case v1beta1.TaskRunResultType:
-			v := v1beta1.ResultValue{}
-			err := v.UnmarshalJSON([]byte(r.Value))
-			if err != nil {
-				continue
-			}
-			taskRunResult := v1beta1.TaskRunResult{
-				Name:  r.Key,
-				Type:  v1beta1.ResultsType(v.Type),
-				Value: v,
+			taskRunResult := v1beta1.TaskRunResult{}
+			if neededTypes[r.Key] == string(v1beta1.ResultsTypeString) {
+				taskRunResult = v1beta1.TaskRunResult{
+					Name:  r.Key,
+					Type:  v1beta1.ResultsTypeString,
+					Value: *v1beta1.NewStructuredValues(r.Value),
+				}
+			} else {
+				v := v1beta1.ResultValue{}
+				err := v.UnmarshalJSON([]byte(r.Value))
+				if err != nil {
+					continue
+				}
+				taskRunResult = v1beta1.TaskRunResult{
+					Name:  r.Key,
+					Type:  v1beta1.ResultsType(v.Type),
+					Value: v,
+				}
```

Comment:
It's interesting to see that UnmarshalJSON never returns an error. For example, if one writes an invalid json string to an object result, UnmarshalJSON will not return error, and instead set the result type to string, and return nil. This will cause the task run to fail because of `missmatched Types for these results` - declared is object but the taskrun  parses as string.

The following taskrun is an example. We do need to return some error and fail the taskrun, but I am curious if the `missmatch` error is desired/appropriate, or we should handle this case differently.
```yaml
apiVersion: tekton.dev/v1beta1
kind: TaskRun
metadata:
  name: obj-result
spec:
  taskSpec:
    results:
      - name: echo-output
        type: object
        properties:
          key: {}
    steps:
      - name: write-invalid-object
        image: bash
        script: |
          set -e
          echo -n '{"key": "value"' | tee $(results.echo-output.path)
```

Reply:
yes very good point!
I think the previous problem is that we we do UnmarshalJSON we don't know what the type we want to Unmarshal to, so if it is an invalid json, it should work if the expected type is string (no error return).

But I think we may improve this, since in this PR we pass the results type to this function. And we may use the type to guide the UnmarshalJSON